packages/gatsby/src/utils/webpack/static-query-mapper.ts

```diff
@@ -0,0 +1,213 @@
+      ) {
+        continue
+      }
+
+      webpackModulesByStaticQueryId.set(id, webpackModule)
+    }
+
+    for (const [id, component] of components) {
+      const componentComponentPath = getRealPath(
+        realPathCache,
+        component.componentPath
+      )
+      if (!doesModuleMatchResourcePath(componentComponentPath, webpackModule)) {
+        continue
+      }
+
+      webpackModulesByComponentId.set(id, webpackModule)
+    }
+  })
+
+  return { webpackModulesByStaticQueryId, webpackModulesByComponentId }
+}
+
+export class StaticQueryMapper {
+  private store: Store<IGatsbyState>
+  private name: string
+
+  constructor(store) {
+    this.store = store
+    this.name = `StaticQueryMapper`
+  }
+
+  apply(compiler: Compiler): void {
+    const { components, staticQueryComponents } = this.store.getState()
+
+    compiler.hooks.done.tap(this.name, stats => {
+      const compilation = stats.compilation
+      // We only care about the main compilation
+      // Chunkgraph should alwasy be available when it's done but you know typescript.
+      if (compilation.compiler.parentCompilation || !compilation.chunkGraph) {
+        return
+      }
+
+      const staticQueriesByChunkGroup = new Map<ChunkGroup, Array<string>>()
+      const chunkGroupsWithPageComponents = new Set<ChunkGroup>()
+      const chunkGroupsByComponentPath = new Map<
+        IGatsbyPageComponent["componentPath"],
+        ChunkGroup
+      >()
+
+      const {
+        webpackModulesByStaticQueryId,
+        webpackModulesByComponentId,
+      } = getWebpackModulesByResourcePaths(
+        compilation.modules,
+        staticQueryComponents,
+        components
+      )
+
+      // group hashes by chunkGroup for ease of use
+      for (const [
+        staticQueryId,
+        webpackModule,
+      ] of webpackModulesByStaticQueryId) {
+        for (const chunk of compilation.chunkGraph.getModuleChunksIterable(
+          webpackModule
+        )) {
+          for (const chunkGroup of chunk.groupsIterable) {
+            const staticQueryHashes =
+              staticQueriesByChunkGroup.get(chunkGroup) ?? []
+
+            staticQueryHashes.push(
+              (staticQueryComponents.get(
+                staticQueryId
+              ) as IGatsbyStaticQueryComponents).hash
+            )
+
+            staticQueriesByChunkGroup.set(chunkGroup, staticQueryHashes)
+          }
+        }
+      }
+
+      // group chunkGroups by componentPaths for ease of use
+      for (const [
+        componentPath,
+        webpackModule,
+      ] of webpackModulesByComponentId) {
+        for (const chunk of compilation.chunkGraph.getModuleChunksIterable(
+          webpackModule
+        )) {
+          for (const chunkGroup of chunk.groupsIterable) {
+            chunkGroupsWithPageComponents.add(chunkGroup)
+            chunkGroupsByComponentPath.set(componentPath, chunkGroup)
+          }
+        }
+      }
```

Comment:
won't this add static queries that are in chunks created by dynamic import (for example) to `globalStaticQueries`?

Reply:
does it also mean that `gatsby develop` and `gatsby build` will have different results because of different chunk settings?

Reply:
I moved to async requires so build & develop are using the same chunkGroups and calculate staticQueries correctly in both cases.